auto_round/quantizers/base.py

```diff
@@ -0,0 +1,1298 @@
+# Copyright (c) 2025 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import copy
+import functools
+import os
+import re
+import sys
+import traceback
+from dataclasses import asdict
+from enum import IntEnum
+from typing import Any, Callable, Dict, List, Union
+
+import accelerate
+import torch
+from accelerate.big_modeling import dispatch_model, infer_auto_device_map
+from transformers import set_seed
+
+from auto_round.data_type import QUANT_FUNC_WITH_DTYPE
+from auto_round.schemes import PRESET_SCHEMES, QuantizationScheme, preset_name_to_scheme
+from auto_round.special_model_handler import _handle_moe_model
+from auto_round.utils import (
+    INNER_SUPPORTED_LAYER_TYPES,
+    SUPPORTED_DTYPES,
+    SUPPORTED_FORMATS,
+    SUPPORTED_LAYER_TYPES,
+    TORCH_VERSION_AT_LEAST_2_6,
+    CpuInfo,
+    _gguf_args_check,
+    _is_fp8_linear,
+    _is_fp8_model,
+    block_forward,
+    check_and_mark_fp8_model,
+    check_is_cpu,
+    check_need_act_calibration,
+    check_seqlen_compatible,
+    check_skippable_keywords,
+    check_to_quantized,
+    clear_memory,
+    collect_best_params,
+    compile_func,
+    convert_dtype_str2torch,
+    convert_fp8_layer_to_linear,
+    convert_fp8_model_to_16b_model,
+    detect_device,
+    find_matching_blocks,
+    flatten_list,
+    get_block_names,
+    get_layer_config_by_gguf_format,
+    get_layer_features,
+    get_layer_names_in_block,
+    get_lm_head_name,
+    get_max_vram,
+    get_module,
+    get_quant_keys,
+    get_shared_keys,
+    htcore,
+    infer_bits_by_data_type,
+    init_cache,
+    is_debug_mode,
+    is_mx_fp,
+    is_nv_fp,
+    is_optimum_habana_available,
+    is_standard_fp,
+    is_static_wfp8afp8,
+    is_wfp8afp8,
+    llm_load_model,
+    logger,
+    mv_module_from_gpu,
+    reset_params,
+    set_amax_for_all_moe_layers,
+    set_module,
+    to_device,
+    to_dtype,
+    unsupport_meta_device,
+)
+
+
+class QuantizerType(IntEnum):
+    MODE = 1
+    MODEL_TYPE = 2
+    DATA_TYPE = 3
+
+
+class BaseQuantizer:
```

Comment:
It might be clearer to start with just the class definitions and add implementation details later.
If possible, a diagram illustrating the call flow and the relationship between `BaseQuantizer` and the other components would also be helpful.